### PR TITLE
Loop unrolling of WildCopy.…

### PR DESCRIPTION
**Pull request by redknightlois**
20% improvement on Decompression
40% improvement on Compression.
cc @jbayardo

### CODE REVIEW

File: src/Sparrow/Compression/LZ4.cs [hunk 1]
```diff
@@ -793,24 +793,26 @@ private static TResult ThrowException<TResult>(Exception e)
...
+            do
+            {
+                ((ulong*)dest)[0] = ((ulong*)src)[0];
+                if (dest + sizeof(ulong) >= destEnd)
+                    return;
 
-            byte* dest = destPtr;
-            byte* src = srcPtr;
-            byte* destEnd = destEndPtr;
+                ((ulong*)dest)[1] = ((ulong*)src)[1];
```
Comment (ayende) [inline src/Sparrow/Compression/LZ4.cs:804]: What about the JIT copy?

File: src/Sparrow/Compression/LZ4.cs [hunk 2]
```diff
@@ -793,24 +793,26 @@ private static TResult ThrowException<TResult>(Exception e)
...
+                if (dest + 2 * sizeof(ulong) >= destEnd)
+                    return;
 
-            // This copy will use the same data that has already being copied as source
-            // It is more of a repeater than a copy per-se. 
+                ((ulong*)dest)[2] = ((ulong*)src)[2];
+                if (dest + 3 * sizeof(ulong) >= destEnd)
```
Comment (ayende) [inline src/Sparrow/Compression/LZ4.cs:809]: Isn't this a lot of conditionals that we can skip? What if we did it in groups of four, and then group of 1 in the end?

**Review by ayende — APPROVED** [review]

File: src/Sparrow/Compression/LZ4.cs (hunk 2 above)
Comment (redknightlois) [reply src/Sparrow/Compression/LZ4.cs:809]: Nope. They cannot be done that way. If we dont check we are guaranteed to touch memory which is out of bounds, and you dont want to :D

File: src/Sparrow/Compression/LZ4.cs (hunk 1 above)
Comment (redknightlois) [reply src/Sparrow/Compression/LZ4.cs:804]: Problem is, this is not a copy... It is a repeater (mascareding as copy). Changing the behavior implies trying to find if we have overlaps and that incurr cost we cannot afford to pay (we went -2x trying to do that).

**Pull request merged**